codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServiceConfigGenerator.kt

```diff
@@ -125,20 +243,201 @@ class ServiceConfigGenerator(
+                }
+                
+                #{BuilderBuildMethod:W}
+            }
+            """,
+            *codegenScope,
+            "BuilderRequiredMethodFlagsInit" to builderRequiredMethodFlagsInit(),
+            "BuilderRequiredMethodFlagDefinitions" to builderRequiredMethodFlagsDefinitions(),
+            "BuilderRequiredMethodError" to builderRequiredMethodError(),
+            "InjectedMethods" to injectedMethods(),
+            "BuilderRequiredMethodFlagsMove1" to builderRequiredMethodFlagsMove(),
+            "BuilderRequiredMethodFlagsMove2" to builderRequiredMethodFlagsMove(),
+            "BuilderRequiredMethodFlagsMove3" to builderRequiredMethodFlagsMove(),
+            "BuilderBuildMethod" to builderBuildMethod(),
+        )
+    }
+
+    private fun isBuilderFallible() = configMethods.isBuilderFallible()
+
+    private fun builderBuildRequiredMethodChecks() = configMethods.filter { it.isRequired }.map {
+        writable {
+            rust(
+                """
+                if !self.${it.requiredBuilderFlagName()} {
+                    return Err(${serviceName}ConfigError::${it.requiredErrorVariant()});
+                }
+                """,
+            )
+        }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsDefinitions() = configMethods.filter { it.isRequired }.map {
+        writable { rust("pub(crate) ${it.requiredBuilderFlagName()}: bool,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsInit() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: false,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsMove() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: self.${it.requiredBuilderFlagName()},") }
+    }.join("\n")
+
+    private fun builderRequiredMethodError() = writable {
+        val variants = configMethods.filter { it.isRequired }.map {
+            writable {
+                rust(
+                    """
+                    ##[error("service is not fully configured; invoke `${it.requiredBuilderFlagName()}` on the config builder")]
+                    ${it.requiredErrorVariant()},
+                    """,
+                )
+            }
+        }
+        if (isBuilderFallible()) {
+            rustTemplate(
+                """
+                ##[derive(Debug, #{ThisError}::Error)]
+                pub enum ${serviceName}ConfigError {
+                    #{Variants:W}
+                }
+                """,
+                "ThisError" to ServerCargoDependency.ThisError.toType(),
+                "Variants" to variants.join("\n"),
+            )
+        }
+    }
+
+    private fun injectedMethods() = configMethods.map {
+        writable {
+            val paramBindings = it.params.map { binding ->
+                writable { rustTemplate("${binding.name}: #{BindingTy},", "BindingTy" to binding.ty) }
```

Comment:
Will this automatically include any dependencies required by `BindingTy`?

Reply:
Yes.